### PR TITLE
21: Implement performance metrics

### PR DESCRIPTION
**Pull request by TobiasAnzinger**
Add calculation of pearson correlation which can be enabled by using the
 flag '--measure_performance pearson'
Add documentation in README.md
Fix numpy version issues by specifying numpy version <2.0

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+
+# Tensorboard
+By running `tensorboard --logdir=runs` a local instance of TensorBoard will be started (http://localhost:6007/)
```
Comment (jualat) [inline README.md:3]:
```suggestion
By running `tensorboard --logdir=runs` a local instance of TensorBoard will be started (http://localhost:6006/)
```
Minor remark: for most users it will be port `6006` and `6007` if the other one is in use

File: framework/performance_metrics.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        if len(self.predictions) == 0 or len(self.ground_truths) == 0:
+            raise ValueError("No data available to compute Pearson correlation.")
+        if len(self.predictions) != len(self.ground_truths):
+            raise ValueError("Mismatch in the length of predictions and ground truths.")
+
+        dev_predictions = self.predictions - np.mean(self.predictions)
+        dev_ground_truths = self.ground_truths - np.mean(self.ground_truths)
+
+        numerator = np.sum(dev_predictions * dev_ground_truths)
+        denominator = np.sqrt(np.sum(dev_predictions ** 2) * np.sum(dev_ground_truths ** 2))
```
Comment (jualat) [inline framework/performance_metrics.py:49]: Looks good :+1:

File: framework/sac_rlhf.py [hunk 3]
```diff
@@ -489,6 +501,15 @@ def get_action(self, x):
...
                     writer.add_scalar("losses/alpha_loss", alpha_loss.item(), global_step)
+                if args.measure_performance == "pearson":
+                    metrics.add_rewards(
+                        np.concatenate(reward_net(data.observations, data.actions).detach().cpu().numpy(), axis=0),
+                        np.concatenate(data.rewards.detach().cpu().numpy(), axis=0)
+                    )
+                    pearson_corr = metrics.compute_pearson_correlation()
+                    writer.add_scalar("charts/pearson_correlation", pearson_corr, global_step)
+                    print(f"Pearson Correlation: {pearson_corr}")
+                    # metrics.reset()
```
Comment (jualat) [inline framework/sac_rlhf.py:512]: I would prefer to not comment this, this could lead to weird behaviour in the unlikely case one wants to try it this metric somewhere else in the code. Even if this does not happen, I see no downside of just doing this always.
But I see the point, so do it as you like.

We should keep in mind that the meaning of reward changes by merging #18

**Review by jualat — APPROVED** [review]

**Comment by jualat** [comment]
#28 is merged now, see comment above

**Comment by Wizzardexe** [comment]
Please update your implementation to align with our current implementation.

**Review by Wizzardexe — CHANGES_REQUESTED** [review]
Solve merge conflict, solve linter issue and recheck if it aligns with our current main branch.

**Review by Wizzardexe — APPROVED** [review]

**Review by bkuen — APPROVED** [review]

**Review by Wizzardexe — APPROVED** [review]
